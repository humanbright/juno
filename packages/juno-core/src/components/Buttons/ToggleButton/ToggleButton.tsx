import React, {
  ComponentProps,
  forwardRef,
  useLayoutEffect,
  useMemo,
  useRef,
} from 'react';

import MuiToggleButton from '@material-ui/lab/ToggleButton';

import {
  combineClasses,
  RcBaseProps,
  removeClassName,
  styled,
  UnionPick,
  useForkRef,
  useThemeProps,
} from '../../../foundation';
import { RcIcon, RcIconProps } from '../../Icon';
import { withTooltip } from '../../Tooltip';
import { RcIconButtonProps } from '../IconButton';
import { ToggleButtonStyle } from './styles';
import { RcToggleButtonClasses } from './utils';

type RcToggleButtonVariant = 'standard' | 'box';

type RcToggleButtonProps = {
  /** variant type for different display `ToggleButton` */
  variant?: RcToggleButtonVariant;
  /** IconProps apply on symbol */
  IconProps?: RcBaseProps<RcIconProps, 'symbol'>;
  /** button size, with default with `medium`   */
  size?: UnionPick<
    NonNullable<RcIconButtonProps['size']>,
    'small' | 'medium' | 'large'
  >;
} & Pick<RcIconButtonProps, 'symbol'> &
  RcBaseProps<ComponentProps<typeof MuiToggleButton>, 'size'>;

const _RcToggleButton = forwardRef<any, RcToggleButtonProps>(
  (inProps: RcToggleButtonProps, ref) => {
    const props = useThemeProps({ props: inProps, name: 'RcToggleButton' });
    const {
      classes: classesProp,
      children,
      size,
      variant,
      symbol,
      IconProps,
      ...rest
    } = props;
    const innerRef = useRef<HTMLButtonElement>(null);

    const buttonRef = useForkRef(ref, innerRef);

    const classes = useMemo(
      () => combineClasses(RcToggleButtonClasses, classesProp),
      [classesProp],
    );

    // * need remove not need Mui-selected for render correct
    useLayoutEffect(() => {
      removeClassName(innerRef, 'Mui-selected');
    });

    return (
      <MuiToggleButton
        {...rest}
        ref={buttonRef}
        innerRef={buttonRef}
        classes={classes}
      >
        {symbol && <RcIcon symbol={symbol} size="inherit" {...IconProps} />}
        {children}
      </MuiToggleButton>
    );
  },
);

const RcToggleButton = styled(withTooltip(_RcToggleButton))`
  ${ToggleButtonStyle}
`;

RcToggleButton.defaultProps = {
  disableRipple: true,
  size: 'medium',
  variant: 'standard',
};

RcToggleButton.displayName = 'RcToggleButton';

export { RcToggleButton };
export type { RcToggleButtonProps, RcToggleButtonVariant };
