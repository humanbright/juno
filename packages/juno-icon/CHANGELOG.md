
## [1.9.0](https://github.com/ringcentral/juno/compare/juno-icon-v1.8.0...juno-icon-v1.9.0) (2022-06-21)


### Features

* **FIJI-52441:** [Icon] add icon `JumpTo` ([bdc8123](https://github.com/ringcentral/juno/commit/bdc81238038cc031078f222cf3b63821bdb6528d))

## [1.8.0](https://github.com/ringcentral/juno/compare/juno-icon-v1.7.0...juno-icon-v1.8.0) (2022-06-13)


### Features

* **FIJI-51927:** [Icon] add icons: AudioLow, TodayCalendarIco ([ff9a48f](https://github.com/ringcentral/juno/commit/ff9a48fa051790070121bc84e019a4db874efe7d))

## [1.7.0](https://github.com/ringcentral/juno/compare/juno-icon-v1.6.1...juno-icon-v1.7.0) (2022-05-12)


### Features

* **FIJI-50567:** [Icon] add icon HybridContact, update icon AppsCategories, AppsDiscovery, AppsProduct ([c8bd630](https://github.com/ringcentral/juno/commit/c8bd630d0e355ccacd83151f46b756d5e6d6dbc8))

### [1.6.1](https://github.com/ringcentral/juno/compare/juno-icon-v1.6.0...juno-icon-v1.6.1) (2022-05-05)


### Bug Fixes

* **update-icon-220505:** [Icon] fix size of AppsInstalled ([e6d4cc0](https://github.com/ringcentral/juno/commit/e6d4cc0e078ff5e969f430e1ad5430d984e7b9d3))

## [1.6.0](https://github.com/ringcentral/juno/compare/juno-icon-v1.5.0...juno-icon-v1.6.0) (2022-05-05)


### Features

* **update-icon-220505:** [Icon] add icons AppsCategories, AppsDevelopers, AppsDiscovery, AppsInstalled, AppsProduct ([889d824](https://github.com/ringcentral/juno/commit/889d8248a3609f2a74ea07e66081b55709eeabb0))

## [1.5.0](https://github.com/ringcentral/juno/compare/juno-icon-v1.4.0...juno-icon-v1.5.0) (2022-04-28)


### Features

* **FIJI-50006:** [Icon] add icon `shield` ([b255589](https://github.com/ringcentral/juno/commit/b255589e6cbb7a590c38a221432719f557db3535))

### [1.2.1](https://github.com/ringcentral/juno/compare/juno-icon-v1.2.0...juno-icon-v1.2.1) (2022-04-02)


### Bug Fixes

* **update-icon-220402:** [icon] fix icon `record_border` ([9dc7428](https://github.com/ringcentral/juno/commit/9dc74283d7d2b4cd27800ffd83c7251b469890bf))

## [1.1.0](https://github.com/ringcentral/juno/compare/juno-icon-v1.0.3...juno-icon-v1.1.0) (2022-03-17)


### Features

* **FIJI-48126:** [Icon] remove icon `box_border`; update icon `box`; add icon `voicemail_border`, `record_border` ([3152943](https://github.com/ringcentral/juno/commit/3152943032d1c50c6940cd687add13b943ae70c7))

### [1.0.3](https://github.com/ringcentral/juno/compare/juno-icon-v1.0.2...juno-icon-v1.0.3) (2022-03-08)


### Features

* **update-icon-220308:** [Icon] update `Customize` ([5ee9851](https://github.com/ringcentral/juno/commit/5ee9851d0afea20cfe29698b8e282935497f2c1f))

### [1.0.2](https://github.com/ringcentral/juno/compare/juno-icon-v1.0.1...juno-icon-v1.0.2) (2022-03-07)


### Features

* **update-icon-220307:** [Icon] add icon ParkLocation, ParkLocationFull; update icon AvatarRooms ([2409f3f](https://github.com/ringcentral/juno/commit/2409f3f6af389420f7709b9af85ebb6163847099))

### [1.0.1](https://github.com/ringcentral/juno/compare/juno-icon-v1.0.0...juno-icon-v1.0.1) (2022-02-23)


### Features

* **juno-icon:** update-icon ([e741a8d](https://github.com/ringcentral/juno/commit/e741a8db4ebe5ca28e3a30cca93d3bd4c768f1fe))
