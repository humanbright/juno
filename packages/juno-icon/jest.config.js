const baseConfig = require('../../jest-base.config');

module.exports = {
  ...baseConfig,
  moduleNameMapper: {
    '@ringcentral/juno$': '<rootDir>/../juno-core/index.ts',
    '@ringcentral/juno-foundation$':
      '<rootDir>/../juno-core/src/foundation/index.ts',
  },
};
