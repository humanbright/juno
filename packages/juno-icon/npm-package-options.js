module.exports = {
  movePaths: ['./README.md', { from: './assets', to: './' }],
};
