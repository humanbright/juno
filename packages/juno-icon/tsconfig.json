{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "allowJs": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true
  },
  "files": [],
  "include": []
}
