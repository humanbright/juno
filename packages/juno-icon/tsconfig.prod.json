{
  "extends": "../../tsconfig.prod.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "./src"
  },
  "include": ["./src/**/*.tsx", "./src/**/*.ts"]
}
