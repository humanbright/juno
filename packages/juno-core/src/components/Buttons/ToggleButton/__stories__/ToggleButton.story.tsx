import React, { ComponentProps } from 'react';

import { Meta, Story } from '@storybook/react';

import { AlignJustify } from '@ringcentral/juno-icon';
import {
  notControlInDocTable,
  notShowInDocTable,
  sortInDocTable,
  switchToControlKnobs,
  Title,
} from '@ringcentral/juno-storybook';

import { RcToggleButton } from '../ToggleButton';

export default {
  title: 'ðŸš€ Cleanup Components/Buttons/ToggleButtonGroup/ToggleButton',
  component: RcToggleButton,
  argTypes: {
    ...sortInDocTable<keyof ToggleButtonProps>([
      'variant',
      'size',
      'symbol',
      'IconProps',
      'selected',
      'disabled',
    ]),
    ...notControlInDocTable<keyof ToggleButtonProps>([]),
    ...notShowInDocTable<keyof ToggleButtonProps>([]),
  },
} as Meta;

type ToggleButtonProps = ComponentProps<typeof RcToggleButton>;

export const ToggleButton: Story<ToggleButtonProps> = ({
  children,
  ...args
}) => {
  switchToControlKnobs();
  return (
    <>
      <Title>ToggleButton should use with RcToggleButton</Title>
      <RcToggleButton value="justify" {...args} symbol={AlignJustify} />
    </>
  );
};

ToggleButton.storyName = 'ToggleButton';

ToggleButton.args = {};

ToggleButton.argTypes = {
  ...notControlInDocTable<keyof ToggleButtonProps>([]),
};

ToggleButton.parameters = {
  tags: [
    {
      name: 'Spec',
      href: 'https://app.abstract.com/projects/d893ad32-0d8d-4e65-a37a-b4a5c881c51b/branches/3c7760bb-8d54-4def-9b30-a50638e618db/commits/4bed494768a18c9e3d10d11dec38590203d12fd3/files/8ffaf9bc-6939-4ef9-9bd1-6aa5b460e2ae/layers/450E7402-70DD-4134-9F26-27B08F55459B?mode=design',
    },
    {
      name: 'Mui',
      href: 'https://material-ui.com/components/toggle-button/#toggle-buttons',
    },
    {
      name: 'Accessibility',
      value: '100%',
    },
  ],
};
