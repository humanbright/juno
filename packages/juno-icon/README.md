# Juno-icon

### Install

```ts
npm i @ringcentral/juno-icon
// or
yarn add @ringcentral/juno-icon
```

## Use Icon

We can use any component with `RcIcon` prop `symbol`, pass component into symbol that will render svg icon with our `RcIcon` style.

```tsx
import { RcIcon } from '@ringcentral/juno';

import { Phone } from '@ringcentral/juno-icon';

const App = (props) => {
  return (
    <RcThemeProvider>
      <RcIcon symbol={Phone} />
    </RcThemeProvider>
  );
};
```
