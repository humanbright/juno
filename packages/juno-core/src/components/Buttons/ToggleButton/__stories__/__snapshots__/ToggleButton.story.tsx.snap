// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ðŸš€ Cleanup Components/Buttons/ToggleButtonGroup/ToggleButton ToggleButton 1`] = `
Array [
  .c0 {
  font-size: 1.5rem;
  font-weight: 400;
  font-family: Lato,Helvetica,Arial,sans-serif;
  line-height: 32px;
  color: #121212;
}

@media (hover:hover) and (pointer:fine) {

}

@media (hover:hover) and (pointer:fine) {

}

<h5
    className="MuiTypography-root c0 MuiTypography-paragraph MuiTypography-displayBlock"
    data-color="neutral.f06"
    data-variant="headline1"
  >
    ToggleButton should use with RcToggleButton
  </h5>,
  .c2 {
  width: 1em;
  height: 1em;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
  pointer-events: none;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: inherit;
}

.c0 {
  color: #666666;
  border: unset;
  padding: 4px;
}

.c0.RcToggleButton-selected {
  color: #066FAC;
}

.c0.RcToggleButton-disabled {
  color: #939393;
}

.c0 .RcToggleButton-label {
  position: relative;
  font-size: 20px;
}

.c0.focus-visible:after,
.c0[data-focus-visible-added]:after {
  content: '';
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  box-shadow: inset 0 0 0 1px #066FAC;
  border-radius: 2px;
  pointer-events: none;
}

.c0:active {
  background-color: rgba(102,102,102,0.12);
}

.c0.RcToggleButton-selected,
.c0.RcToggleButton-selected:hover {
  background-color: rgba(6,111,172,0.12);
}

@media (hover:hover) and (pointer:fine) {
  .c0:hover {
    color: #121212;
  }
}

@media (hover:hover) and (pointer:fine) {
  .c0:hover {
    background-color: rgba(102,102,102,0.08);
  }
}

<button
    className="MuiButtonBase-root MuiToggleButton-root c0"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
    value="justify"
  >
    <span
      className="MuiToggleButton-label RcToggleButton-label"
    >
      <span
        className="c1 align-justify icon"
      >
        <svg
          className="c2"
          viewBox="0 0 32 32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M27 24a1 1 0 0 1 0 2H5a1 1 0 0 1 0-2h22zm0-9a1 1 0 0 1 0 2H5a1 1 0 0 1 0-2h22zm0-9a1 1 0 0 1 0 2H5a1 1 0 0 1 0-2h22z"
          />
        </svg>
        
      </span>
    </span>
  </button>,
]
`;
